### PR TITLE
MSQ window function: Take segment granularity into consideration to fix NPE issues with ingestion

### PR DESCRIPTION
**Pull request by Akshat-Jain**
### Description

Currently, MSQ queries like the following failed with a NPE:
```sql
REPLACE INTO "test_1" OVERWRITE ALL
SELECT
  "__time",
  "countryName",
  ROW_NUMBER() over (PARTITION BY countryName) as c1
FROM "wikipedia"
where countryName='Guatemala'
PARTITIONED BY DAY
```

The root cause was that the shuffle spec of the final window stage was being hardcoded to MixShuffleSpec, hence it wasn't taking the required segment granularity into account. This PR changes the logic to use the `resultShuffleSpecFactory` for the last window stage.

Along with the above, we also need to add virtual columns like `__bucket` to the frame writer in `WindowOperatorQueryFrameProcessor`. This PR makes this change as well.

This PR also adds a test which fails with NPE without the changes of this PR.

<hr>

This PR has:

- [x] been self-reviewed.
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java [hunk 1]
```diff
@@ -119,6 +125,20 @@ public WindowOperatorQueryFrameProcessor(
...
+    frameWriterVirtualColumns.add(partitionBoostVirtualColumn);
+
+    final VirtualColumn segmentGranularityVirtualColumn =
+        QueryKitUtils.makeSegmentGranularityVirtualColumn(jsonMapper, query);
+
+    if (segmentGranularityVirtualColumn != null) {
+      frameWriterVirtualColumns.add(segmentGranularityVirtualColumn);
+    }
+
+    this.frameWriterVirtualColumns = VirtualColumns.create(frameWriterVirtualColumns);
```
Comment (kgyrtkirk) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java:141]: I think this should be remove from here as it seems like a cross cutting aspect
The same lines are present in other `QueryKit` classes

could you refactor these changes out from all `QueryKit` classes and place them at a common place?

**Review by kgyrtkirk — COMMENTED** [review]
this is again an issue arising from the fact that the last shuffle/etc should be different if its being stored as files

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java [hunk 2]
```diff
@@ -105,8 +108,20 @@ public QueryDefinition makeQueryDefinition(
 
     final int firstStageNumber = Math.max(minStageNumber, queryDefBuilder.getNextStageNumber());
     final WindowOperatorQuery queryToRun = (WindowOperatorQuery) originalQuery.withDataSource(dataSourcePlan.getNewDataSource());
-    final int maxRowsMaterialized;
 
+    // Get segment granularity from query context, and create ShuffleSpec and RowSignature to be used for the final window stage.
```
[the remark this answers is not in the document]
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:112]: We already have the redundant logic abstracted away in `QueryKitUtils.getSegmentGranularityFromContext()` etc methods. Rest of the logic doesn't seem common across all the query kits to me.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java (hunk 1 above)
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java:141]: These lines are same in `ScanQueryFrameProcessor`, but they are different in `GroupByPostShuffleFrameProcessor`, and not present in `GroupByPreShuffleFrameProcessor`. So I don't see how to cleanly create a base class for use by all frame processors.

Also, it feels unrelated to this PR changes, so we should tackle it separately in my opinion. Would like to keep the changes affect only the window-related layers if possible.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kgyrtkirk) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:112]: I think there should be a better way

As far as I understand thing here this approach exposes an execution time complexity and applies them to all stages
however it should only affect the last stage - and as such it should be naturally there in the system and only apply it to the last stage (regardless what type of query is being run)

[the remark this answers is not in the document]
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:112]: Can you elaborate? We are using the resultShuffleSpecFactory only for the final window stage, not for all of them.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java [hunk 3]
```diff
@@ -122,13 +137,13 @@ public QueryDefinition makeQueryDefinition(
       queryDefBuilder.add(
           StageDefinition.builder(firstStageNumber)
                          .inputs(new StageInputSpec(firstStageNumber - 1))
-                         .signature(rowSignature)
+                         .signature(finalWindowStageRowSignature)
                          .maxWorkerCount(maxWorkerCount)
-                         .shuffleSpec(null)
+                         .shuffleSpec(finalWindowStageShuffleSpec)
```
[the remark this answers is not in the document]
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:142]: This is being done only for the last window stage. If we have an empty over() clause present in the window function, we create a single window stage instead of separate stages.

With that said, I've pushed a change to add the virtual columns also only for the last window stage. Hope that works!

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:112]: Have pushed a change to add the virtual columns also only for the last window stage. Hope that works!

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:142]: From what I can understand, you are adding logic which uses the segment granularity to decide the shuffling. However, it would be possible to have more than one window stage in a query, and the query does not need to be followed by a segment generation stage. You can add a limit to a window query to see this behavior. I am wondering if this case is covered as well.

[the remark this answers is not in the document]
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:142]: @adarshsanjeev We're using the resultShuffleSpecFactory only for the last window stage. I'm not sure I fully understand the comment.

For the final stage, we have the following cases for segmentGranularity:
1. Granularities.ALL: We set the shuffle spec to MixShuffleSpec for the final window stage in this case.
2. All other types: We create the shuffle spec using resultShuffleSpecFactory for the final window stage in this case.

I tried running the following query as well, as suggested (IIUC):
```sql
select countryName, cityName, channel, 
row_number() over (partition by countryName order by countryName, cityName, channel) as c1, 
row_number() over (partition by cityName order by countryName, cityName, channel) as c2
from wikipedia
where countryName in ('Republic of Korea', 'Austria')
group by countryName, cityName, channel
limit 7
```

This query gives correct results.

[the remark this answers is not in the document]
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:142]: For the above query, these are the stages we get:
```
stageDefinition.getProcessorFactory() = org.apache.druid.msq.querykit.groupby.GroupByPreShuffleFrameProcessorFactory@737a033
stageDefinition.getSignature() = {d0:STRING, d1:STRING, d2:STRING}
stageDefinition.getStageNumber() = 0
stageDefinition.getShuffleSpec() = MaxCountShuffleSpec{clusterBy=ClusterBy{columns=[d0 ASCENDING, d1 ASCENDING, d2 ASCENDING], bucketByCount=0}, partitions=1, aggregate=true}


stageDefinition.getProcessorFactory() = org.apache.druid.msq.querykit.groupby.GroupByPostShuffleFrameProcessorFactory@3b66dec
stageDefinition.getSignature() = {d0:STRING, d1:STRING, d2:STRING}
stageDefinition.getStageNumber() = 1
stageDefinition.getShuffleSpec() = HashShuffleSpec{clusterBy=ClusterBy{columns=[d0 ASCENDING], bucketByCount=0}, numPartitions=1}


stageDefinition.getProcessorFactory() = org.apache.druid.msq.querykit.WindowOperatorQueryFrameProcessorFactory@d45add62
stageDefinition.getSignature() = {d1:STRING, d0:STRING, d2:STRING, w0:LONG}
stageDefinition.getStageNumber() = 2
stageDefinition.getShuffleSpec() = HashShuffleSpec{clusterBy=ClusterBy{columns=[d1 ASCENDING], bucketByCount=0}, numPartitions=1}


stageDefinition.getProcessorFactory() = org.apache.druid.msq.querykit.WindowOperatorQueryFrameProcessorFactory@643913d9
stageDefinition.getSignature() = {d0:STRING, d1:STRING, d2:STRING, w0:LONG, w1:LONG}
stageDefinition.getStageNumber() = 3
stageDefinition.getShuffleSpec() = MixShuffleSpec{}


stageDefinition.getProcessorFactory() = org.apache.druid.msq.querykit.scan.ScanQueryFrameProcessorFactory@687279fb
stageDefinition.getSignature() = {__boost:LONG, d0:STRING, d1:STRING, d2:STRING, w0:LONG, w1:LONG}
stageDefinition.getStageNumber() = 4
stageDefinition.getShuffleSpec() = MixShuffleSpec{}


stageDefinition.getProcessorFactory() = org.apache.druid.msq.querykit.common.OffsetLimitFrameProcessorFactory@3c8
stageDefinition.getSignature() = {__boost:LONG, d0:STRING, d1:STRING, d2:STRING, w0:LONG, w1:LONG}
stageDefinition.getStageNumber() = 5
stageDefinition.getShuffleSpec() = MaxCountShuffleSpec{clusterBy=ClusterBy{columns=[__boost ASCENDING], bucketByCount=0}, partitions=1, aggregate=false}
```

[the remark this answers is not in the document]
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:142]: I meant with INSERT, but I think that case should also work. I wasn't sure initially.

**Review by adarshsanjeev — APPROVED** [review]

**Review by LakshSingla — CHANGES_REQUESTED** [review]
**Note:** This comment is only for the changes made to the frame processor and the factory. I don't have enough context on the window query kit to approve the changes there, but I will try to leave some questions based on my understanding. 

IIUC this patch only adds virtual columns conditionally. We can add them unconditionally, and modify the signature to include that column, and that's a better way to avoid setting the flag.

For example, if the `ColumnSelectorFactory` that's input to a frame writer has valid `ColumnValueSelectors` for columns a, b, c, however the [signature with which the frame writer factory is created](https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/frame/write/RowBasedFrameWriterFactory.java#L47-L47) is a, b, d, then the written frame will have the following format:

| a | b | d |
|--------|--------|--------|
| val1a | val1b | * |
| va2a | val2b | * | 

Depending on the implementation of the `ColumnSelectorFactory`, if it returns nil-selector for a missing column, the column  d will be populated with nulls, else if the factory throws then the frame signature will throw while writing. Column c doesn't have any relevance in the final frame. 

All this is to say that addition of a column to `ColumnSelectorFactory` doesn't determine the frame's columns (and thereby it's signature), its the signature that was created when creating the FrameWriterFactory, i.e. the input to the FrameWriterFactory#makeRowBasedFrameWriterFactory. 

Here are a couple of things that can be made better with this patch:

1. We are adding a partition boosting virtual column to the `ColumnSelectorFactory`, while the signature with which we are writing the frame doesn't have this boost column. This boosted column isn't doing anything. To materialize the boost column in the frame, we should add the boost column in the corresponding signature with which this frame writer is created. But, I would leave that change for another patch, and simply cleanup the boosting virtual column that has been added, since it isn't doing anything
2. The `segmentGranularityVirtualColumn` can be added unconditionally in the `WindowOperatorQueryFrameProcessor`. It doesn't need to be gated behind any flag. Since only the last window requires that column, adding the virtual column shouldn't make any difference for "not-last" windows. This should remove the need for any modifications for the frame processor factories, which are actually communicated b/w the worker/controller, and any changes must be vetted for upgrade/downgrade compatibility.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java [hunk 4]
```diff
@@ -119,6 +126,21 @@ public WindowOperatorQueryFrameProcessor(
     for (int i = 0; i < frameReader.signature().size(); i++) {
       typeStrategies[i] = frameReader.signature().getColumnType(i).get().getNullableStrategy();
     }
+
+    // Get virtual columns to be added to the frame writer.
+    this.partitionBoostVirtualColumn = new SettableLongVirtualColumn(QueryKitUtils.PARTITION_BOOST_COLUMN);
+    final List<VirtualColumn> frameWriterVirtualColumns = new ArrayList<>();
+    if (addVirtualColumns) {
+      frameWriterVirtualColumns.add(partitionBoostVirtualColumn);
```
Comment (LakshSingla) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java:134]: Boosting isn't done since the query kit isn't setting the boost column. The boost column should be explicitly added in the window query kit for this line to take effect. We can take it up in a separate PR.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java [hunk 5]
```diff
@@ -119,6 +126,21 @@ public WindowOperatorQueryFrameProcessor(
...
+
+      final VirtualColumn segmentGranularityVirtualColumn =
+          QueryKitUtils.makeSegmentGranularityVirtualColumn(jsonMapper, query);
+
+      if (segmentGranularityVirtualColumn != null) {
+        frameWriterVirtualColumns.add(segmentGranularityVirtualColumn);
```
Comment (LakshSingla) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java:140]: Likewise, this needn't be gated behind `addVirtualColumns` flag. Last window would have set the signature correctly (which is why it is getting used), and the other stages wouldn't have used it anyways (otherwise you'd have seen the error earlier).

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java [hunk 6]
```diff
@@ -99,7 +105,8 @@ public WindowOperatorQueryFrameProcessor(
       final List<OperatorFactory> operatorFactoryList,
       final RowSignature rowSignature,
       final int maxRowsMaterializedInWindow,
-      final List<String> partitionColumnNames
+      final List<String> partitionColumnNames,
+      final boolean addVirtualColumns
```
Comment (LakshSingla) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java:109]: Can revert this change. LMK if you face any error - correctness or otherwise while running the query after making the suggested changes.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessorFactory.java
Comment (LakshSingla) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessorFactory.java:1]: We can revert the changes in this file since we don't wanna rely on a flag.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java (hunk 4 above)
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java]: Have removed the relevant lines.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java (hunk 5 above)
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java]: Makes sense, have made the change!

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java (hunk 6 above)
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java]: Done

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessorFactory.java
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessorFactory.java:1]: Done

**Comment by Akshat-Jain** [comment]
@LakshSingla Thanks a lot for the detailed explanation, it cleared a bunch of things for me. Have made the suggested changes, please have a look. Thanks! 😄

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java [hunk 7]
```diff
@@ -119,6 +123,15 @@ public WindowOperatorQueryFrameProcessor(
...
     }
+
+    // Get virtual columns to be added to the frame writer.
+    final List<VirtualColumn> frameWriterVirtualColumns = new ArrayList<>();
+    final VirtualColumn segmentGranularityVirtualColumn =
+        QueryKitUtils.makeSegmentGranularityVirtualColumn(jsonMapper, query);
+    if (segmentGranularityVirtualColumn != null) {
+      frameWriterVirtualColumns.add(segmentGranularityVirtualColumn);
+    }
+    this.frameWriterVirtualColumns = VirtualColumns.create(frameWriterVirtualColumns);
```
[the remark this answers is not in the document]
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java]: Done

[the remark this answers is not in the document]
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java:134]: @LakshSingla Had to revert the change. A bunch of tests started failing because `VirtualColumns#create` doesn't work with null values, it throws NPE.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java [hunk 8]
```diff
@@ -105,8 +108,20 @@ public QueryDefinition makeQueryDefinition(
...
     final WindowOperatorQuery queryToRun = (WindowOperatorQuery) originalQuery.withDataSource(dataSourcePlan.getNewDataSource());
-    final int maxRowsMaterialized;
 
+    // Get segment granularity from query context, and create ShuffleSpec and RowSignature to be used for the final window stage.
+    final Granularity segmentGranularity = QueryKitUtils.getSegmentGranularityFromContext(jsonMapper, queryToRun.getContext());
+    final ClusterBy finalWindowClusterBy = QueryKitUtils.clusterByWithSegmentGranularity(ClusterBy.none(), segmentGranularity);
+    ShuffleSpec finalWindowStageShuffleSpec = resultShuffleSpecFactory.build(finalWindowClusterBy, false);
+    if (Objects.equals(segmentGranularity, Granularities.ALL)) {
+      finalWindowStageShuffleSpec = MixShuffleSpec.instance();
+    }
```
Comment (LakshSingla) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:118]: This seems wrong. We should still be using the `resultShuffleSpecFactory` to create the final shuffle spec. If the windowing stage is an input to the segment generator, then the window processor would create one large partition, while we expect the segments to be sized correctly. Am I missing the reason for this additional check?

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java [hunk 9]
```diff
@@ -105,8 +108,20 @@ public QueryDefinition makeQueryDefinition(
 
     final int firstStageNumber = Math.max(minStageNumber, queryDefBuilder.getNextStageNumber());
     final WindowOperatorQuery queryToRun = (WindowOperatorQuery) originalQuery.withDataSource(dataSourcePlan.getNewDataSource());
-    final int maxRowsMaterialized;
 
+    // Get segment granularity from query context, and create ShuffleSpec and RowSignature to be used for the final window stage.
+    final Granularity segmentGranularity = QueryKitUtils.getSegmentGranularityFromContext(jsonMapper, queryToRun.getContext());
+    final ClusterBy finalWindowClusterBy = QueryKitUtils.clusterByWithSegmentGranularity(ClusterBy.none(), segmentGranularity);
```
Comment (LakshSingla) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:114]: Why is this `ClusterBy.none()`? I am not certain if that's wrong, but we should mention a reason as to why the final stage doesn't sort by anything apart from time.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java
[the remark this answers is not in the document]
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:1]: @LakshSingla 
1. Fair enough. I added them for debuggability purposes. Since it's not anything excessive, I was thinking we can tone these down once we've had a bit of soak time for these flows (as major changes are going into these code flows currently).
2. Fair enough. There's a pending refactoring work item as well. I wanted to add exhaustive tests along with that. Can also add exhaustive documentation with that. Would certainly want to take it up separate from this PR though, if that works.
3. Fair enough. Similar as point (2).
4. Fair enough. Similar as point (2).
    1. Yes we can have multiple sort/partition operators to the input. Their ordering is done in `Windowing.java` file, and MSQ receives them properly ordered. We can potentially improve in that regards (wrt further optimisations), but that's outside the scope of the MSQ specific files. Hope that answers.
    2. Yeah it's correct. If we don't have any partitioning requirements, then we don't need to do any shuffling, and hence any sorting if needed can happen via NaiveSortOperatorFactory itself during processing.
    3. "What if a partition isn't present in the sort columns." --- can you please elaborate what you mean by that?

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java (hunk 8 above)
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:118]:
```java
    if (Objects.equals(segmentGranularity, Granularities.ALL)) {
      finalWindowStageShuffleSpec = MixShuffleSpec.instance();
    }
```

Removing the above lines causes failures when we have multiple workers (`maxNumTasks` > 2).

Queries fail with:
```
Caused by: org.apache.druid.java.util.common.IAE: Cannot shuffle with spec [MaxCountShuffleSpec{clusterBy=ClusterBy{columns=[], bucketByCount=0}, partitions=4, aggregate=false}] and nil clusterBy
```

Thoughts?

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java (hunk 9 above)
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:114]: I'm not sure what ClusterBy should be used here. I don't see any ClusterBy to extract from the available fields in this class. Thoughts? Please let me know if I'm missing something. Thanks!

Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java]: I've pushed some logic changes as per offline discussion. Please have a look, will close this comment as it's outdated now. Thanks!

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java (hunk 8 above)
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java]: I've pushed some logic changes as per offline discussion. Please have a look, will close this comment as it's outdated now. Thanks!

**Comment by adarshsanjeev** [comment]
The changes look good to me % remaining Laksh's comments

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java [hunk 10]
```diff
@@ -360,4 +354,36 @@ private QueryDefinitionBuilder makeQueryDefinitionBuilder(String queryId, DataSo
     }
     return queryDefBuilder;
   }
+
+  /**
+   * Computes the ClusterBy for the final window stage which may or may not have the partition boosted column,
+   * depending on the {@code segmentGranularity} parameter passed. We don't have to take the CLUSTERED BY
+   * columns into account, as they are handled as {@link org.apache.druid.query.scan.ScanQuery#orderBys}.
+   */
+  private static ClusterBy computeClusterByForFinalWindowStage(Granularity segmentGranularity)
```
Comment (LakshSingla) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:363]: Partition boosting should be done irrespective of the granularity IMO.

Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:363]: Why though?

Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:363]: What's the harm?

Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:363]: It's an unnecessary column, so we shouldn't add it unless we need to? Same reason why we are not adding it to all the window stages.

Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:363]: Hmm, I see your point.

Why is partition boosting conditional in GroupByQueryKit then?

Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:363]: Currently, the boost column is a workaround for when the clustered by is empty (i.e. granularity = ALL). It is also an optimization that breaks up large partitions. Hence even if the query doesn't throw when granularity =/= ALL, it still is worth it to have that optimisation for other granularities as well.

Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:363]: @LakshSingla IIUC, do you mean it should be unconditional in GroupByQueryKit as well?

Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:363]: GroupBy is different due to the fact that it aggregates as well, therefore the chance of having a large partition if the partition is on the grouping dimensions is negligible (haven't seen anything like that in practice). Therefore there can't be large partitions when the post shuffle spec = pre shuffle spec. There can be large partitions if that isn't true which is when we have group by + order by etc. Window, scan kits don't aggregate, therefore they have to worry about partition sizes.

Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java:360]: @LakshSingla I see. Have made the suggested change to add partition boosting irrespective of segment granularity. Thanks for pointing this out!

**Review by LakshSingla — APPROVED** [review]
